drivers/usb/device/usb_dc_nrfx.c

```diff
@@ -1375,13 +1375,14 @@ int usb_dc_ep_write(const u8_t ep, const u8_t *const data,
 	}
 
 
-	int sem_status = k_sem_take(&ctx->dma_in_use, K_NO_WAIT);
+	k_sem_take(&ctx->dma_in_use, K_FOREVER);
 
-	/* USBD hardware does not allow scheduling multiple DMA transfers
-	 * at a time. Next USB transfer can be triggered after all DMA
-	 * operations are complete.
+	/* USBD driver does not allow scheduling multiple DMA transfers
+	 * for one EP at a time. Next USB transfer on this endpoint can be
+	 * triggered after the completion of previous one.
 	 */
-	if (sem_status != 0) {
+	if (ep_ctx->write_in_progress) {
```

Comment:
The `sem_give` seems to be used as a lock in this file (so why semaphore now mutex?) and `write_in_progress` seems to be protected by it (in this function).

The `usbd_event_handler` is called without `dma_in_used` taken if I am correct. There is a race between here and that handler function. It seems however that we don't care.
Note that `usbd_event_handler` can set `write_in_progress` (even when volatile) in a way that you don't intent it to do. If you depend on this set to occur in specific place (e.g. after or before something) please add barrier. You can use atomic or simply lock access to the variable using the same lock as here.

I suggest adding a comment that everything is fine because:
- we don't care about the race between `usbd_event_handler` and this function and,
- locking `dma_in_progress` ensures that `write_is_progress` will me set to true by only one thread.